napari/_qt/qt_color_dialog.py

```diff
@@ -0,0 +1,267 @@
+import re
+from typing import Optional
+
+import numpy as np
+from qtpy.QtCore import Qt, Signal, Slot
+from qtpy.QtGui import QColor
+from qtpy.QtWidgets import (
+    QColorDialog,
+    QCompleter,
+    QFrame,
+    QHBoxLayout,
+    QLineEdit,
+    QVBoxLayout,
+    QWidget,
+)
+
+from ..layers.utils.color_transformations import ColorType
+from ..utils.colormaps.standardize_color import (
+    hex_to_name,
+    rgb_to_hex,
+    transform_color,
+)
+from .qt_modal import QtPopup
+from vispy.color import get_color_dict
+
+# matches any 3- or 4-tuple of int or float, with or without parens
+# captures the numbers into groups.
+# this is used to allow users to enter colors as e.g.: "(1, 0.7, 0)"
+rgba_regex = re.compile(
+    r"\(?([\d.]+),\s*([\d.]+),\s*([\d.]+),?\s*([\d.]+)?\)?"
+)
+
+TRANSPARENT = np.array([0, 0, 0, 0], np.float32)
+UNKNOWN = -1  # a constant to indicate that the user entered an invalid string
+FALLBACK = TRANSPARENT  # what will be shown if a user entry is UNKNOWN
+
+
+class QColorSwatchEdit(QWidget):
+    """A widget that combines a QColorSwatch with a QColorLineEdit.
+
+    emits a color_changed event with a 1x4 numpy array when the current color
+    changes.  Note, the "model" for the current color is the ``_color``
+    attribute on the QColorSwatch.
+    """
+
+    color_changed = Signal(np.ndarray)
+
+    def __init__(
+        self,
+        parent: Optional[QWidget] = None,
+        *,
+        initial_color: Optional[ColorType] = None,
+        tooltip: Optional[str] = None,
+    ):
+        """Create a new ColorSwatchEdit widget.
+
+        Parameters
+        ----------
+        parent : QWidget, optional
+            parent widget, by default None
+        initial_color : ColorType, optional
+            Starting color, by default None
+        tooltip : str, optional
+            Tooltip when hovering on the swatch,
+            by default 'click to set color'
+        """
+        super().__init__(parent=parent)
+        self.setObjectName('QColorSwatchEdit')
+
+        layout = QHBoxLayout()
+        layout.setContentsMargins(0, 0, 0, 0)
+        layout.setSpacing(6)
+        self.setLayout(layout)
+
+        self.lineEdit = QColorLineEdit(self)
+        self.lineEdit.editingFinished.connect(self._on_lineedit_edited)
+
+        self.colorSwatch = QColorSwatch(self, tooltip=tooltip)
+        self.colorSwatch.color_changed.connect(self._on_swatch_changed)
+        self.setColor = self.colorSwatch.setColor
+        if initial_color is not None:
+            self.setColor(initial_color)
+
+        layout.addWidget(self.colorSwatch)
+        layout.addWidget(self.lineEdit)
+
+    def color(self):
+        """Return the current color."""
+        return self.colorSwatch.color()
+
+    def _on_lineedit_edited(self):
+        """When the user hits enter or loses focus on the LineEdit widget."""
+        text = self.lineEdit.text()
+        rgb_match = rgba_regex.match(text)
```

Comment:
Just so I understand if this where the autocomplete functionality is happening? I've not used autocomplete in Qt before, could you just say a little something about where / how that is working (and say why I don't seem to get transparent to autocomplete)

Reply:
ah I see it below - sorry I should be doing this in review mode!!